src/CacheManager.Core/BaseCacheManager.Keys.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using CacheManager.Core.Logging;
+
+namespace CacheManager.Core
+{
+    public sealed partial class BaseCacheManager<TCacheValue>
+    {
+        /// <inheritdoc />
+        override public IEnumerable<string> Keys(string pattern, string region)
+        {
+            CheckDisposed();
+
+            var keys = new HashSet<string>();
+
+            if (_logTrace)
+            {
+                Logger.LogTrace("Keys [{0}:{1}] started.", region, pattern);
+            }
+
+            for (var handleIndex = 0; handleIndex < _cacheHandles.Length; handleIndex++)
+            {
+                var handle = _cacheHandles[handleIndex];
+                var handleKeys = handle.Keys(pattern ?? "*", region);
+                keys.UnionWith(handleKeys);
+            }
+
+            if (_logTrace)
+            {
+                Logger.LogTrace("Keys [{0}:{1}] found {2} keys.", region, pattern, keys.Count);
+            }
+
+            if (region != null)
+            {
+                var after = region.Length + 1; // +1 for the :
+                return keys.Select(k => k.Substring(after));
```

Comment:
Thing is that the cache handles do have specific logic for regions and eventually an instance token, too. So the full key could be {instanceToken}@{region}@{key} or only {region}:{key} or just {key}.

That's why I meant that the cache handle itself would have to implement the logic to substract all prefixes from the key and return the key only. 
The central logic in BaseCacheManager cannot do this